src/runtime/thread_pool_common.h

```diff
@@ -79,109 +159,383 @@ struct work_queue_t {
+    const char *parent_name = job->parent_job ? (job->parent_job->task.name ? job->parent_job->task.name : "<no name>") : "<no parent job>";
+    log_message(prefix << name << "[" << job << "] serial: " << job->task.serial << " active_workers: " << job->active_workers << " min: " << job->task.min << " extent: " << job->task.extent << " siblings: " << job->siblings << " sibling count: " << job->sibling_count << " min_threads " << job->task.min_threads << " next_sempaphore: " << job->next_semaphore << " threads_reserved: " << job->threads_reserved << " parent_job: " << parent_name << "[" << job->parent_job << "]");
+    for (int i = 0; i < job->task.num_semaphores; i++) {
+        log_message(indent << "    semaphore " << (void *)job->task.semaphores[i].semaphore << " count " << job->task.semaphores[i].count << " val " << *(int *)job->task.semaphores[i].semaphore);
     }
-    return desired_num_threads;
 }
 
-WEAK int default_desired_num_threads() {
-    int desired_num_threads = 0;
-    char *threads_str = getenv("HL_NUM_THREADS");
-    if (!threads_str) {
-        // Legacy name for HL_NUM_THREADS
-        threads_str = getenv("HL_NUMTHREADS");
+WEAK void dump_job_state() {
+    log_message("Dumping job state, jobs in queue:");
+    work *job = work_queue.jobs;
+    while (job != NULL) {
+        print_job(job, "    ");
+        job = job->next_job;
     }
-    if (threads_str) {
-        desired_num_threads = atoi(threads_str);
-    } else {
-        desired_num_threads = halide_host_cpu_count();
-    }
-    return desired_num_threads;
+    log_message("Done dumping job state.");
 }
+#else
+#define print_job(job, indent, prefix)
+#define dump_job_state()
+#endif
+
+WEAK void worker_thread(void *);
 
 WEAK void worker_thread_already_locked(work *owned_job) {
-    // If I'm a job owner, then I was the thread that called
-    // do_par_for, and I should only stay in this function until my
-    // job is complete. If I'm a lowly worker thread, I should stay in
-    // this function as long as the work queue is running.
-    while (owned_job != NULL ? owned_job->running()
-           : work_queue.running()) {
-
-        if (work_queue.jobs == NULL) {
+    while (owned_job ? owned_job->running() : !work_queue.shutdown) {
+        work *job = work_queue.jobs;
+        work **prev_ptr = &work_queue.jobs;
+
+        if (owned_job && owned_job->exit_status != 0 && owned_job->active_workers == 0) {
+            while (job != owned_job) {
+                prev_ptr = &job->next_job;
+                job = job->next_job;
+            }
+            *prev_ptr = job->next_job;
+            job->task.extent = 0;
+            continue; // So loop exit is always in the same place.
+        }
+
+        dump_job_state();
+
+        // Find a job to run, prefering things near the top of the stack.
+        while (job) {
+            print_job(job, "", "Considering job ");
+            // Only schedule tasks with enough free worker threads
```

Comment:
Is this comment still a sufficient description? This sort of thing has been changing.